### ISSUE TITLE
Add contentMediaType to GeoJSON example

### ISSUE THREAD

**Issue by jerstlouis** [opengeospatial/ogcapi-processes#223]
During the workshop, it was found that we could not easily detect outputs available as GeoJSON, due to `application/geo+json` not appearing anywhere in output descriptions. GeoJSON is represented as `"type" : "object"`, and JSONSchema says:

> The _contentMediaType_ keyword specifies the MIME type of the contents of a **string**, as described in RFC 2046. 

However, nothing seems to preclude its use with objects in addition to string.

It was also discussed that although schemas can be included e.g. `"$ref" : "https://geojson.org/schema/FeatureCollection.json"`, different schemas can be specified for GeoJSON if specific types of objects are expected (e.g. https://geojson.org/schema/MultiPolygon.json), and that this inclusion essentially requires parsing the content of those schema and is equivalent to directly inlining the schema's content. This would make it extremely difficult to recognize GeoJSON by schema only.

Therefore, we recommend to add `"contentMediaType" : "application/geo+json"` to the OGC process description/input descriptions examples where GeoJSON is returned, and maybe add some clarifications to that effect.

**Comment by pvretano** [comment]
@jerstlouis, @gfenoy pointed to this link: https://json-schema.org/understanding-json-schema/reference/non_json_data.html
It seems to imply that contentMediaType is meant to be used with strings.

There is also this link: https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.8.4 which again indicates that contentMediaType is for string so I don't think me can merge the PR associated with this issue.

I suppose the following is valid:
```
{
   "type": "string",
   "contentSchema": "https://geojson.org/schema/FeatureCollection.json",
   "contentMediaType": "application/geo+json"
}
```
@jerstlouis @gfenoy  what do you think?

**Comment by jerstlouis** [comment]
@pvretano I had seen all this (At the workshop, I posted a comment on the gitter with those exact links when I brought this up), and that is what was saying again today: JSON Schema did intend this to be used with strings only.

`"type" : "strying"  as  you just put it would mean the whole JSON string is encoded within double quotes, not as an embedded JSON object, so that doesn't work.

I still think we really badly need a mechanism to identify GeoJSON content embedded as an object as "application/geo+json".

So this is why I suggest that if this "extension" of JSONSchema (using _contentMediaType_ in another context) does not cause validation to fail with the JSON Schema meta-schema, I still strongly recommend that we go ahead with adding this to the example as well as a recommendation.

**Comment by pvretano** [comment]
@jerstlouis I would not say we "badly" need this.  JSON is not as strongly bound to schema as say XML so if I put a GeoJSON feature as the value of the input, any JSON parser would be able to parse it.  My JSON parser detects if the JSON object is GeoJSON by looking for the "type", "geometry" and "property" keys.  If those exist and the value of "type" is "FeatureCollection" or "Feature" then it assumes it is GeoJSON.
Looking at [https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/schema.yaml](schema.url)  I don't think it will cause validation issues BUT I am not keen about recommending something that clearly goes against what the JSON schema specification says.

**Comment by jerstlouis** [comment]
@pvretano in our client implementation, we have a dedicated module to loading GeoJSON and we wouldn't even make it to that module if we don't know off-hand to expect GeoJSON.
Our JSON parser (used inside that module) also requires a type before even trying to parse something.

That would mean a lot of ugly work-arounds to try to determine beforehand if it is GeoJSON we're parsing this, and since the JSON object could be any of thousands of different things, that really doesn't work.

**Comment by pvretano** [comment]
@jerstlouis the crux of the matter is that we need to convey, semantically, that this JSON object is a GeoJSON object.  I took a look at [https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7](JSON Schema) and I believe that the correct way to handle this situation is:
```
{
   "$ref": "http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/geometryGeoJSON.yaml",
   "format": "geojson"      
}
```
So, the only thing I would need to do is add a small section to the specification defining the "geojson" tag as a valid format tag as per section [https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.2.3](7.2.3) of the JSON specification.  I would also add a small informative explanation and example.

NOTE: We CANNOT use contentMediaType because that key is specifically for strings as per the JSON Schema specification.

@jerstlouis @gfenoy is this OK with you?

**Comment by jerstlouis** [comment]
@pvretano I would be happy if `format` is the way to go, however this needs to handle more than GeoJSON.
Could we specify that this should be the media type when using `"type" : "object"`, so that it scales to every JSON format for which a media type is defined? Or otherwise provide an extension mechanism to other potential JSON formats?

**Comment by gfenoy** [comment]
I would like to point this reference: https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-14.2.

I think it may be related in that sense that a featureCollection is an instance of the a schema (defining the featureCollection, [FeatureCollection.json](https://geojson.org/schema/FeatureCollection.json)).

**Comment by pvretano** [comment]
@gfenoy I think this is just the media type for a JSON schema file.  No?  This still does not tell us that the semantic content of the schema is, say, GeoJSON.

@jerstlouis none of the existing format values are media types so I assuming that we should not use media types.  Based on my understanding of the JSON schema section i cited, we can either:

1. Reuse the `format` key and define additional values like `geojson`.  We could also perhaps say that additional values can be added to the format list using the format "ogc-x-<value>".
2. Using the recommendation in last paragraph of section [7.2.3](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.2.3) and define a new key to add to the schema.yaml file, say we call it `mediaType`, for the purpose for pointing to additional semantic information about the type.  So instead of format in my example above we would say `"mediaType": "application/geo+json"`

@jerstlouis @gfenoy let me know what you think.

**Comment by jerstlouis** [comment]
@pvretano The place I see `format` is still for strings, e.g. https://json-schema.org/understanding-json-schema/reference/string.html#format
And nowhere in https://json-schema.org/understanding-json-schema/reference/object.html

Approach 1 -- that would be fine, but where would implementers find the list of those OGC-defined values, and what would be the process to register new ones? That's why re-using IANA media types make sense.

Approach 2 -- that sounds good. However that is awfully similar to `contentMediaType`. Couldn't we just extend it to `mediaType`? I guess that would conflict with the one typically used for strings... I don't really mind if we call it `contentMediaType`, `mediaType` or `objectMediaType`. The important thing is to be able to look at a process description and easily find out that you can retrieve the output as GeoJSON, or any other potential format.

**Comment by pvretano** [comment]
@jerstlouis we cannot use "contentMediaType" since JSON schema already says this is a facet for strings and not objects.  I don't mind "objectMediaType" but I feel it would be better to refer to an existing structure like `format` since that does not require us to extend the structure of [schema.yaml](https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/schema.yaml).

 I am not sure what the validation impact of `objectMediaType` would be.  I guess non-OGC-API-Processes schema parsers would just ignore it.

@gfenoy your thoughts?

**Comment by gfenoy** [comment]
I do like the `objectMediaType` idea but, I do feel more comfortable with the approach 1 also and agree about OGC NA.

Also, sorry for going back on the previous comment about [section-14.2](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-14.2.) but, the format may be only the string: `geo` (the `schema-instance` here is geo in my understanding) in the example you provided where we used: `"mediaType": "application/geo+json"`.

We may also use  `x-ogc-geo` as you explained, nevertheless, I was understanding that "x-" was applied for spefic keys rather than for value but, I am fine with it too.

**Comment by pvretano** [comment]
@gfenoy  I suppose saying `"format": "geo"` is OK but don't you think `"format": "geojson"` is more clear?  I would propose the following list of format values: `geojson-feature-collection`, `geojson-feature`, `geojson-geometry` and `ogc-bbox` to represent the common JSON input types: a GeoJSON  feature collection, a GeoJSON feature, a GeoJSON geometry and an OGC bbox (i.e. bbox.yaml).

@jerstlouis you OK with option 1 keeping in mind that we will:

1. define an extension mechanism (i.e. ogc-x-<value>)
2. Use the OGC NA to manage the additional format tags.

**Comment by jerstlouis** [comment]
@pvretano I am okay with any of that. But I feel like finding a way to use of any IANA-defined media type is going to be much easier for implementer using some media type no one registered yet and not wanting to bother registering it / delays in this happening / extra work on OGC NA / etc.

**Comment by pvretano** [comment]
@gfenoy sorry, I think I misunderstood what you are saying.  You are saying that the value of the `format` key can only be the schema-instance value from the media type ... correct?  Where specifically does it say that in the JSON schema specifications?  I don't think that Section [7.2.3](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.2.3) says that.

**Comment by jerstlouis** [comment]
@pvretano I guess an even stronger argument for using the exact media type is because the input description lists the valid values that can be used in the execution request outputs / format key.

**Comment by gfenoy** [comment]
@pvretano, I was trying to make a parallel in between the following:

`"mediaType": "application/geo+json"`
and
`"mediaType": "application/schema-instance+json"` (from section [4.2](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-00#section-14.2))

We know that we will always have an object, it is obviously an "application/json" per default. Nevertheless, we are willing to refine the `mediaType`. In consequence, I would like to have the `format` (which would probably benefit from being named `instanceOf` for clarity) to specify only the missing information, in such a case `geo`.  

After reading again the discussion here, I feel this is not exactly what you are talking about. So I feel a bit off-topic finally. Nevertheless, I don't see why we should add mediaType information when referencing one schema from [here](http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/) should be enough. We have already all the informations we require from: [featureCollectionGeoJSON.yaml](http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/featureCollectionGeoJSON.yaml) / [FeatureCollection.json](https://geojson.org/schema/FeatureCollection.json),   [featureGeoJSON.yaml](http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/featureGeoJSON.yaml) / [Feature.json](https://geojson.org/schema/Feature.json) and finally, [geometryGeoJSON.yaml](http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/geometryGeoJSON.yaml) / [Geometry.json](https://geojson.org/schema/Geometry.json) (corresponding to the  values you were willing to add to format except for the Bbox). 

So, IMO nothing really need to be changed here. Still, I think that `"instanceOf": "geojson-feature-collection"` is a reasonable value (or any of the values @pvretano was referring to: `geojson-feature`, `geojson-geometry` and `ogc-bbox`) for making it easier for server to understand what is requested without relying on the schema reference that may change in time. As it is only a recommendation and not a requirement, I would encourage adding `instanceOf` or `format` to simplify the refinement of input/output type definitions.

**Comment by pvretano** [comment]
@gfenoy is `instanceOf` an official JSON schema key?  ... or are you proposing we add that?

**Comment by gfenoy** [comment]
I am proposing we add that but, format is fine too but, I do not like to use similar term for different things usually.

**Comment by jerstlouis** [comment]
@gfenoy Those schemas do not specify the media type though, and the only way for the client to determine that `application/geo+json` is a valid value for the output's format's mediaType is to do an in-depth analysis of the schema to determine that it is in fact, GeoJSON. And it has to recognize *all* of these schemas as GeoJSON.
This is way to much to ask of the client, in my opinion.

**Comment by pvretano** [comment]
@gfenoy I agree with you about reusing similar terms for different things but if I am reading the discussion is [7.2.3](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.2.3) correctly, it says

_Structural validation alone may be insufficient to allow an application to correctly utilize certain values. The "format" annotation keyword is defined to allow schema authors to convey semantic information for a fixed subset of values which are accurately described by authoritative resources, be they RFCs or other external specifications.  The value of this keyword is called a format attribute._ 

So I believe the `format` key is OK since we are looking for a semantic hint to tell us that the schema is in-fact GeoJSON (in the case of the example in the specification).

If there are no objections then, I will add a recommendation that say to use `format` and I will define the following values:

- geojson-feature-collection
- geojson-feature
- geojson-geometry
- ogc-bbox

and allows extensions via a `ogc-x-<format tag>` tag.

**Comment by jerstlouis** [comment]
@pvretano If we go with this approach, there should probably be a column in the registered table listing the media type, which is what should be used in the execute request to specify the media type (or if `Accept:` header are ever supported as in #217).
It also requires the work to set up an OGC NA register of these formats?

**Comment by pvretano** [comment]
@jerstlouis agreed.  I will include a table that has the format value, a reference to a JSON schema and an instance media type.

**Comment by cportele** [comment]
I haven't really followed all the discussion and you are probably aware of this, but just in case: 

Using custom format values is allowed, but it is an option that is discouraged by the JSON Schema spec. See [validation, section 7.2.3](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.7.2.3). 

> Due to this limitation, and the historically uneven implementation of this keyword, it is RECOMMENDED to define additional keywords in a custom vocabulary rather than additional format attributes if interoperability is desired.

There is not much love for "format" among the authors, just have a look at the discussions or release notes. It can be an option for Processes, but it shouldn't be a building block that is recommended for reuse outside of Processes.

**Comment by pvretano** [comment]
@cportele I agree but our concern is that if we add a new key, which I would be happy to do, might cause some existing tools (swagger, etc.) to have issues.  The feeling is that `format`, although discouraged, is known entity and extending its value space is probably safe.  Your thoughts?

**Comment by jerstlouis** [comment]
@cportele do you know whether the use of `contentMediaType` with an object would be a proper Processes extension, or would be discouraged?

**Comment by pvretano** [comment]
@jerstlouis because you don't believe me ... https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#page-17

**Comment by jerstlouis** [comment]
@pvretano I believe you and I read that and I agree that it only mentions string. However it says:

> **If the instance is a string,** this property indicates the media type of the contents of the string.

It doesn't seem to preclude its use with object at all. Which is why I was curious about @cportele 's opinion on the matter.

**Comment by pvretano** [comment]
@jerstlouis ... the title of the section is: "A Vocabulary for the Contents of String-Encoded Data" and it is only mentioned in that section and the "Security Consideratiions" section ... and it that section it is warning about injection attacks ... just the type of thing you do with a STRING type.  If it could be using for other types don't you think they would mention it somewhere else in the specification?

**Comment by pvretano** [comment]
One other related issue ... whether we reuse the `format` key or the `contentMediaType` key or some other key we define I think that the correct schema fragment would need to be (using the example from the specification):
```
        "oneOf": [
          {
            "type": "string",
            "contentMediaType": "application/gml+xml; version=3.2",
            "contentSchema": "http://schemas.opengis.net/gml/3.2.1/geometryBasic2d.xsd"
          },
          {
             "allOf": [
                {
                   "format": "geojson-feature-geometry"
                },
                {
                  "$ref": "http://schemas.opengis.net/ogcapi/features/part1/1.0/openapi/schemas/geometryGeoJSON.yaml"
                }
             ]
          }
        ]
      }
```
In other words, adding this semantic hint would require a schema that is a little more complicated.

**Comment by pvretano** [comment]
@jerstlouis "insist"?  Yes sir! ;)

I'm just grappling with the best approach.  
Reusing the `format` key or defining a new key (either: `instanceOf` or `objectMediaType`).  
As I mentioned above in my response to @cportele I know the authors of the JSON schema specification don't have much love for the `format` key BUT my feeling is that reusing the `format` key is the path of least resistance.  The `format` key is a known entity already defined in JSON schema and all we are doing is extending the value space.  
Adding a new element might have unintended side effects ... like causing some tools (like swagger) to complain.

**Comment by jerstlouis** [comment]
@pvretano sorry, I didn't mean it in that way, just that I still find it super important.

About Swagger, I don't think the process description where this occurs would be used at all with SwaggerUI.

Rather, a Swagger UI would be useful for the "OpenAPI" Process description which would embed the schema, together with the higher level information like multiplicity etc. and the paths above, and there for whichever hint we have (be it `format` or objectMediaType`), the service could include that information (much more easily if it is directly a media type) in the Swagger-defined `contentType` key which sits _above_ the JSON schema.

That would be in the response to the 200 for results or for sync requests. Sadly, making that OpenAPI description of the process work relies very much on the changes proposed #217 I'm afraid.

**Comment by pvretano** [comment]
@jerstlouis, respectfully, I think you need to take a few minutes and review [OGC's Principles of conduct](https://www.ogc.org/ogc/policies/conduct).  I believe that insisting, badgering and obnoxious pushing of your position is outside those principles.  We already discusses #217 in the SWG and I will remind you that I did and do support it.  The SWG, however, decided to table it for now and I think we have to respect that decision and focus on the issue at hand which is the semantic hint.

On that point, I will update the PR and tag you and @gfenoy as reviewers (if I can -- last time, for some reason, I could not).  Please stand by ...

**Comment by jerstlouis** [comment]
@pvretano I apologize, I did not mean to be pushy or _insist on you making a change_, I meant highlighting that I feel this is important. Sorry that it didnot come out as intended.

About #217 I was just pointing out that the whole OpenAPI content negotiation will be an issue without support for `Accept:` header, not that we should. The SWG decided to table it and I accept that.

**Comment by cportele** [comment]
> @cportele I agree but our concern is that if we add a new key, which I would be happy to do, might cause some existing tools (swagger, etc.) to have issues. The feeling is that format, although discouraged, is a known entity and extending its value space is probably safe. Your thoughts?

@pvretano - Yes, I agree, it is a reasonable approach. As long as it is clear that all this is only meant for the very specific design that Processes has decided to use and not recommended for use elsewhere (in particular, in OGC API standards).

> @cportele do you know whether the use of contentMediaType with an object would be a proper Processes extension, or would be discouraged?

@jerstlouis - As @pvretano has said, it is clear that `contentMediaType`/`contentEncoding` in JSON Schema is only meant for non-JSON data encoded in a string value. For JSON-encoded data there is no need for a special construct, the media type of the whole document applies (and includes a sub-object) and no encoding information is necessary. Again, the issue is the result of the Processes design, which supports an edge case of the use of JSON (obviously not from the perspective of an OGC API Processes implementer, otherwise all this wouldn't be in Core, but from the wider developer perspective).

Another pointer: https://json-schema.org/understanding-json-schema/reference/non_json_data.html

> JSON schema has a set of keywords to describe and optionally validate non-JSON data stored inside JSON strings. Since it would be difficult to write validators for many media types, JSON schema validators are not required to validate the contents of JSON strings based on these keywords. However, these keywords are still useful for an application that consumes validated JSON. 
> ...

**Comment by bpross-52n** [comment]
Thanks for the discussion, closed via #230
